lib/cn/csched_sp3_work.c

```diff
@@ -278,25 +285,98 @@ sp3_work_wtype_idle(
+ *    has "split_cnt_max" nodes or fewer they might all be undergoing an
+ *    active spill.
+ *
+ * 3) Defer requesting a split if there are already "split_cnt_max" splits
+ *    active in the tree (i.e., the max number of concurrent splits).
+ *
+ * 4) Defer requesting a split if sufficient time has not passed since
+ *    the first of the last batch of splits started.
+ *
+ * If none of the above conditions hold, then we request a split by setting
+ * the "tn_splitting" flag to true, which will prevent new spills into this
+ * node.  However, if this node is currently undergoing an active spill then
+ * we must return 0 to avoid starting a split.  We then re-evalutate this
+ * condition on each invocation until all spills to this node have completed
+ * (i.e., tn_ss_spilling == 0).  Only then may we start the split.
+ *
+ * The reason for these machinations is to avoid stalling the root spill
+ * pipeline, as that adversely impacts the ingest pipeline, which is
+ * then seen as poor quality-of-service from the client's perspective.
+ */
 static uint
 sp3_work_wtype_split(
     struct sp3_node          *spn,
     struct sp3_thresholds    *thresh,
     struct kvset_list_entry **mark,
     enum cn_action           *action,
-    enum cn_rule             *rule)
+    enum cn_rule             *rule,
+    bool                     *resched)
 {
     struct cn_tree_node *tn = spn2tn(spn);
+    struct cn_tree *tree = tn->tn_tree;
     struct kvset_list_entry *le;
     struct list_head *head;
+    uint64_t now;
+    uint kvsets;
 
     head = &tn->tn_kvset_list;
     *mark = list_last_entry_or_null(head, typeof(*le), le_link);
+    kvsets = cn_ns_kvsets(&tn->tn_ns);
 
     *action = CN_ACTION_SPLIT;
     *rule = CN_RULE_SPLIT;
 
-    return cn_ns_kvsets(&tn->tn_ns);
+    mutex_lock(&tn->tn_ss_lock);
+    now = get_time_ns();
+
+    if (!tn->tn_ss_splitting) {
+        if (tree->ct_split_cnt >= thresh->split_cnt_max || now < tree->ct_split_dly) {
+            tn->tn_ss_visits = 0;
+            kvsets = 0;
+            ev_debug(1);
+        } else if (cn_ns_kvsets(&tree->ct_root->tn_ns) >= thresh->rspill_runlen_min * 3) {
+            tn->tn_ss_visits = 0;
+            kvsets = 0;
+            ev_debug(1);
+        } else if (tn->tn_ss_spilling && tn->tn_ss_visits < thresh->split_cnt_max) {
+            tn->tn_ss_visits++;
+            kvsets = 0;
+            ev_debug(1);
+        } else {
+            if (++tree->ct_split_cnt >= thresh->split_cnt_max)
+                tree->ct_split_dly = now + NSEC_PER_SEC * 3;
+            tn->tn_ss_splitting = true;
+            tn->tn_ss_visits = 0;
+            ev_debug(1);
+        }
+    }
+
+    /* Defer starting the split if there are active spills to this node.
+     */
+    if (tn->tn_ss_spilling)
+        kvsets = 0;
+    mutex_unlock(&tn->tn_ss_lock);
+
+    /* Set the resched flag to prevent csched from dropping this request
+     * should we return 0 to defer the split.
+     */
+    *resched = !kvsets;
```

Comment:
seems like caller can just check for 0 return value instead of having two outputs?

Reply:
I think we're going to need to more logic here to recheck the split condition.  If we only check the return code and don't want to split then csched will just keep asking us over and over unless we can disable it via the resched flag.  I didn't want to overload the return value.  We can reviist once we see how join will work...